File: kNN.c
/*
*	Steven Booth, Nick Goodpaster
*	
*	COEN171 - Exploring Programming Languages Project
*	
*	Description: C implementation of k-Nearest Neighbors
*	machine learning algorithm. Makes a prediction for test 
*	sample given training sample data. Attributes are loaded from
*	iris.csv file, which is split into training and testing data.
*	Attributes include pedal length, pedal width, sepal length, 
*	and sepal width. Identifiers are setosa, virginica, and versicolor. 
*	The program computes a prediction for the testing data with
*	96% accuracy. 
*/


#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <ctype.h>
#define N  150
#define TRAIN 120
#define TEST 30
#define MAX 11

/*	
*	Structure declaration for neighbor type. 
*	Holds ID and distance from test instance
*/
typedef struct{
	float distance;
	char *ID;
} neighbor;

//Function definitions
float distanceFormula(float **train, float *testInstance, int i);
float* getDistances(float **train, float *testInstance);
neighbor* sortDistances(float *distances, char **trainingID);
neighbor* getNN(neighbor *sortedNeighbors, int k);
char* getVote(neighbor *NN, int k);
char* getMax(int setosa, int virginica, int versicolor);
void loadDataSet(char *filename, float **data, char **ID);
float getAccuracy(char **predictions, char **testingID);

//Main function
int main(void){
	int k = 3;
	
	//Allocates memory for testing data set
	float **testingData = (float **) malloc(TEST * sizeof(float *));
  	for(int i = 0; i < TEST; i++){
		testingData[i] = (float *)malloc(4 * sizeof(float));
	}
	char **testingDataID = malloc(TEST * sizeof(char*));
	for(int i = 0; i < TEST; i ++){
		testingDataID[i] = (char *) malloc(10 * sizeof(char));
	}
	
	//Allocates memory for training data set	
	float **trainingData = (float **) malloc(TRAIN * sizeof(float *));
	for(int i = 0; i < TRAIN; i++){
		trainingData[i] = (float *)malloc(4 * sizeof(float));
	}
	char **trainingDataID = malloc(TRAIN * sizeof(char*));
	for(int i = 0; i < TRAIN; i ++){
		trainingDataID[i] = (char *) malloc(10 * sizeof(char));
	}

	//Allocates memory to hold predictions
	char **predictions = malloc(TEST * sizeof(char *));
	for(int i = 0; i < TEST; i ++){
		predictions[i] = (char *) malloc(10 * sizeof(char));
	}
	
	//Loads the data sets from input files into data arrays
	loadDataSet("iristraining.txt",trainingData, trainingDataID);
	loadDataSet("iristesting.txt", testingData, testingDataID);	
	
	//calculates a prediction for each data sample in testingData	
	for(int j = 0; j < TEST; j++){	
		float *distances = getDistances(trainingData, testingData[j]);
		neighbor *neighbors = sortDistances(distances, trainingDataID);	
		neighbor *NN = getNN(neighbors, k); 	
		predictions[j] = getVote(NN, k);	
		printf(">>Prediction: %s   >>>>   Actual: %s \n", predictions[j], testingDataID[j]);
		free(neighbors);
		free(distances);
	}

	//working to fix strcmp in getAccuracy function
	//printf("\n\n the accuracy of the alogirithm is: %f\n", getAccuracy(predictions, testingDataID));
	free(predictions);
	free(testingData);
	free(testingDataID);
	free(trainingData);
	free(trainingDataID);
	return 0;
}

/*
*	Euclidian distance formula, calculates distance between
*	flower attributes. 
*/
float distanceFormula(float **train, float *testInstance, int i){
	float distance = 0;
	distance += pow(testInstance[0] - train[i][0], 2);
	distance += pow(testInstance[1] - train[i][1], 2);
	distance += pow(testInstance[2] - train[i][2], 2);
	distance += pow(testInstance[3] - train[i][3], 2);
	return sqrt(distance);
}

/*
*	Calculates distance between testing instance and every 
*	training value. 
*/
float* getDistances(float **train, float *testInstance){
	//printf("%f, %f\n", testInstance[0], testInstance[1]);
	float *distances = (float *) malloc(sizeof(int) * TRAIN);
	for(int i = 0 ; i < TRAIN; i ++){ 
		float distance = distanceFormula(train, testInstance, i);		
		distances[i] = distance;
	}
	return distances;
}

/*
*	Pairs the distance and the ID to create a neighbor type for each
*	of the distances. Sorts this array of neighbors to be used for 
*	finding the k-Nearest Neighbors.
*/
neighbor* sortDistances(float *distances, char **trainingID){
	neighbor* neighbors = (neighbor *) malloc(TRAIN * sizeof(neighbor));
	neighbor* temp = (neighbor *)  malloc(sizeof(neighbor));
	
	//Loading distances and trainingId's into array of neighbors
	for(int i = 0; i < TRAIN; i ++){
		neighbors[i].distance = distances[i];
		neighbors[i].ID = malloc(10 * sizeof(char));
		strcpy(neighbors[i].ID, trainingID[i]);
	}
	
	//Bubble sort array of neighbors by distance
	for(int i = 0; i < (TRAIN -1); i++){
		for(int j = 0; j < (TRAIN - i - 1); j ++){
			if(neighbors[j].distance > neighbors[j + 1].distance){
				*temp = neighbors[j];
				neighbors[j] = neighbors[j + 1];
				neighbors[j + 1] = *temp;
			}
		}
	}
	return neighbors;			
}

/*
*	Uses the sorted neighbor array to create a subarray containing
*	only the k-Nearest Neighbors of the test instance. 
*/
neighbor* getNN(neighbor *sortedNeighbors, int k){
	neighbor* NN = (neighbor *) malloc (k * sizeof(neighbor));
	for(int i = 0; i < k; i ++){
		NN[i] = sortedNeighbors[i];
	}	
	return NN;
}

/*
*	Each neighbor will vote for it's ID type, these votes will 
*	then be passed into the getMax function to find the majority
*	ID of the neighbors.
*/
char* getVote(neighbor *NN, int k){
        int setosa = 0, virginica = 0, versicolor = 0;
	for(int i = 0; i < k; i ++){
		if(strcmp("setosa", NN[i].ID) == 0){
			setosa++;
		}else if(strcmp("virginica", NN[i].ID) == 0){
			virginica++;	
		}else if(strcmp("versicolor", NN[i].ID) == 0){
			versicolor++;
		}
	}
	return getMax(setosa, virginica, versicolor);
}

/*
*	Returns the prediction, which is the ID with the most 
*	votes. 
*/
char* getMax(int setosa, int virginica, int versicolor){
	char* prediction = malloc(MAX * sizeof(char));
	if(setosa > virginica && setosa > versicolor){
		strcpy(prediction, "setosa");
	} else if(virginica > setosa && virginica > versicolor){
		strcpy(prediction, "virginica");
	} else if(versicolor > setosa && versicolor > virginica){
		strcpy(prediction, "versicolor");
	} else{
		strcpy(prediction, "still cannot be determined");
	}
	return prediction;
}

/*
*	Function to parse the csv files for the training and testing
*	data sets into the testingData and trainingData arrays. 
*/
void loadDataSet(char *filename, float **data, char **ID){
	FILE* file = fopen(filename, "r");
        char line[40];
        char *instance = (char *) malloc(MAX * sizeof(char));
        int i, j, rowcount = 0, IDcount = 0, instancecount = 0;;
        float instanceData;
        while(fgets(line, sizeof(line), file)){
                i = 0;
                instancecount = 0;
                while(i < sizeof(line)){
                        j = 0;
                        if(!isalpha(line[i])){
                                do{
                                        instance[j] = line[i];
                                        j++;
                                        i++;
                                }while(line[i] != ',');
                                instanceData = atof(instance);
                                data[rowcount][instancecount] = instanceData;
                                instance[0] = '\0';
                                instancecount++;
                        } else if(isalpha(line[i])){
                                do{
                                        instance[j] = line[i];
                                        j++;
                                        i++;
                                }while(isalpha(line[i]));
                                instance[j++] = '\0';
                                strcpy(ID[IDcount], instance);
                                IDcount ++;
                                break;
                        }
                        i++;
                }
                rowcount ++;
        }
        free(instance);
	fclose(file);
}

/*
*	Function to calculate the accuracy of the alogorithm by evaluating
*	how many of the testingDataID's were predicted correctly. 
*/

float getAccuracy(char **predictions, char **testingID){
	int correctpredictions = 0;
	for(int i = 0; i < TEST; i ++){
		printf("%s, %s\n", testingID[i], predictions[i]);
		if(strcmp(testingID[i], predictions[i]) == 0){
			correctpredictions++;
		}
	}
	return 100 * (correctpredictions/TEST);
}

File: kNN.py

"""
Steven Booth, Nick Goodpastor

COEN 171 - Exploring Programming Languages Project

Algorithm: k-Nearest Neighbors

Description: This alogirthm will take in the iris.csv file and parse the data 
into training and testing data. Once this is complete, each instance of testing
data will be compared to every training data instance and the distance between
the attributes will be calculated and stored in a distance array. Next, the 
k-nearest neighbors of the test instance will take a vote using their own 
identifiers. The most common identifier amongst the nearest neighbors will be
the prediciton for the test instance. Once all the predictions are made, the 
program will compute its accuracy based on the actual identifers of the testing
data set and the predicted identifiers. Our program runs with a 95% or better
accuracy when using 5 or less neighbors. 

"""

#imported libraries
import math
import csv
import random
import operator

"""
Function to load the data set from the irisdata file and split the data into
training and testing data records. 
"""

def loadDataSet(filename, split, trainingDataSet = [], testingDataSet = []):
    with open(filename, 'r') as csvfile:
        #Reads lines of the csv file using the built in csv reader
        lines = csv.reader(csvfile)
        #makes a list of the lines in the csv file
        data_csv = list(lines)
        for x in range(len(data_csv)):
            #for each of the attributes in the csv file, the values are cast
            #to floats
            for y in range(5):
                data_csv[x][y] = float(data_csv[x][y])
            #creates a random number between 0 and 1.0 and appends value
            #to training or testing data set with a weighted split
            if random.random() < split:
                trainingDataSet.append(data_csv[x])
            else:
                testingDataSet.append(data_csv[x])

"""
Function to compute the distance between a testingInstance and trainingData.

Returns the distance between the testing attribute and the training attribute
"""

def distanceFormula(testingInstance, trainingData, length):
    distance = 0
    for x in range(length):
        distance += pow((float(testingInstance[x])) - (float(trainingData[x])), 2)
    return math.sqrt(distance)
    
    
"""
Function to compute the k-Nearest Neighbors. Takes in a test instance and 
compares the distance between its attributes and the attributes of all the
instance values in the training data. Takes the closest k neighors and returns 
an array containing the k-Nearest Neighbors neighbors. 

Returns the array of nearest neighbors. 
"""
def getNN(trainingDataSet, testInstance, k):
    print('\n')
    #creates an array to hold distances for neighbors
    distances = []
    #sets the length to be the number of attributes, so the number of columns
    #in the test instance minus the identifier for the type of flower
    length = len(testInstance) - 1
    for x in range(len(trainingDataSet)):
        #computes distance between training instance and testing instance
        distance = distanceFormula(testInstance, trainingDataSet[x], length)
        #appends to distances array
        distances.append((trainingDataSet[x], distance))
    #sorts the distances
    distances.sort(key=operator.itemgetter(1))
    neighbors= []
    #appends the k-Nearest neighbors to the neighbors array 
    for x in range(k):
        neighbors.append(distances[x][0])
        print('Neighbor #' + repr(x + 1) + ': ' + repr(neighbors[x][-1]))
    return neighbors
    
"""
Function to get the votes from the nearest neighbors to decide the predicted
type of the testing instance. Will use the getMax function to generate a 
prediction.

Returns string representation of prediction to be added to the predictions 
array in the main. 
"""
def getVotes(neighbors):
    setosaVotes = 0
    virginicaVotes = 0
    versicolorVotes = 0
    for x in range(len(neighbors)):
        if neighbors[x][5] == 'setosa':
            setosaVotes += 1
        if neighbors[x][5] == 'virginica':
            virginicaVotes += 1
        if neighbors[x][5] == 'versicolor':
            versicolorVotes += 1
    prediction = getMax(setosaVotes, virginicaVotes, versicolorVotes)
    return prediction
  
"""
Function to get the maximum value of the three votes and returns the string
representation of the most voted flower type from the neighbors array. 

Returns the prediction in the form of a string. 
"""  

def getMax(setosaVotes, virginicaVotes, versicolorVotes):
    if((setosaVotes > virginicaVotes) & (setosaVotes > versicolorVotes)):
        prediction = 'setosa'
    if((virginicaVotes > setosaVotes) & (virginicaVotes > versicolorVotes)):
        prediction = 'virginica'
    if((versicolorVotes > setosaVotes) & (versicolorVotes > virginicaVotes)):
        prediction = 'versicolor'
    return prediction
          
"""
Function to compute the accuracy of the algorithm by comparing the correct
flower types from the testing data and the predictions computed in our
algorithm.

Returns the value of the accuracy as a float to be printed int the main
function. 
"""
def getAccuracy(testSet, predictions):
    correct = 0
    for x in range(len(testSet)):
        if testSet[x][5] == predictions[x]:
            correct += 1
    return (correct/float(len(testSet))) * 100
    
"""
Main function definition.
"""
def main():
    #prepare data
    trainingDataSet = []
    testingDataSet = []
    split = 0.8
    loadDataSet('iris.csv', split, trainingDataSet, testingDataSet)
    #generate predictions
    predictions=[]
    k = 3
    for x in range(len(testingDataSet)):
        neighbors = getNN(trainingDataSet, testingDataSet[x], k)
        result = getVotes(neighbors)
        predictions.append(result)
        print('Predicted=' + repr(result) + ', Actual=' + repr(testingDataSet[x][-1]))
        print('-------------------------------------------')
    accuracy = getAccuracy(testingDataSet, predictions)
    print('\n\nAlgorithm Accuracy: ' + repr(accuracy) + '%')
    
#main function call     
main()

